File: fall23_preli-B/F.cpp
// Breadth First Search (BFS) - ChatGPT

#include <bits/stdc++.h>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

vector<int> bfs(vector<vector<int>>& graph, int start, int destination) {
    vector<int> parent(graph.size(), -1);
    vector<bool> visited(graph.size(), false);
    queue<int> q;

    q.push(start);
    visited[start] = true;

    while (!q.empty()) {
        int current = q.front();
        q.pop();

        if (current == destination) {
            vector<int> path;
            while (current != -1) {
                path.push_back(current);
                current = parent[current];
            }
            reverse(path.begin(), path.end());
            return path;
        }

        for (int neighbor : graph[current]) {
            if (!visited[neighbor]) {
                q.push(neighbor);
                visited[neighbor] = true;
                parent[neighbor] = current;
            }
        }
    }

    return vector<int>(); // No path found
}

int main() {
    int N, S, D, M;
    cin >> N >> S >> D >> M;

    vector<vector<int>> graph(N + 1);

    for (int i = 0; i < M; ++i) {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
    }

    vector<int> path = bfs(graph, S, D);

    if (path.empty()) {
        cout << "-1" << endl;
    } else {
        cout << path.size() << endl;
        for (int node : path) {
            cout << node << " ";
        }
        cout << endl;
    }

    return 0;
}

File: fall23_preli-B/F.py
# Breadth First Search (BFS) - ChatGPT

from collections import deque

def bfs(graph, start, destination):
    parent = [-1] * len(graph)
    visited = [False] * len(graph)
    queue = deque()

    queue.append(start)
    visited[start] = True

    while queue:
        current = queue.popleft()

        if current == destination:
            path = []
            while current != -1:
                path.append(current)
                current = parent[current]
            return path[::-1]

        for neighbor in graph[current]:
            if not visited[neighbor]:
                queue.append(neighbor)
                visited[neighbor] = True
                parent[neighbor] = current

    return []

def main():
    N, S, D = map(int, input().split())
    M = int(input())

    graph = [[] for _ in range(N + 1)]

    for _ in range(M):
        u, v = map(int, input().split())
        graph[u].append(v)

    path = bfs(graph, S, D)

    if not path:
        print("-1")
    else:
        print(len(path))
        print(*path)

if __name__ == "__main__":
    main()
